integrations/snyk/snyk/client.py

```diff
@@ -0,0 +1,268 @@
+
+        response = await self._send_api_request(
+            url=url, method="GET", query_params=query_params
+        )
+
+        target = response.get("data", {})
+        async for projects_data_of_target in self.get_paginated_projects(
+            target.get("id")
+        ):
+            target.setdefault("__projects", []).extend(projects_data_of_target)
+        return target
+
+    async def get_paginated_targets(
+        self,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        url = f"/orgs/{self.org_id}/targets"
+        query_params = {"version": f"{self.snyk_api_version}~beta"}
+        async for targets in self._get_paginated_resources(
+            base_url=self.rest_api_url, url_path=url, query_params=query_params
+        ):
+            targets_with_project_data = []
+            for target_data in targets:
+                async for projects_data_of_target in self.get_paginated_projects(
+                    target_data.get("id")
+                ):
+                    target_data.setdefault("__projects", []).extend(
+                        projects_data_of_target
+                    )
+                targets_with_project_data.append(target_data)
+            yield targets_with_project_data
+
+    async def get_single_project(self, project: dict[str, Any]) -> dict[str, Any]:
+        url = f"{self.rest_api_url}/orgs/{self.org_id}/projects/{project.get('id')}"
+        query_params = {
+            "version": self.snyk_api_version,
+            "meta.latest_issue_counts": "true",
+            "expand": "target",
+        }
+        response = await self._send_api_request(
+            url=url, method="GET", query_params=query_params
+        )
+        return await self.enrich_project(response.get("data", {}))
+
+    async def get_cached_user_details(self, user_id: str) -> dict[str, Any]:
+        if (
+            not user_id
+        ):  ## Some projects may not have been assigned to any owner yet. In this instance, we can return an empty dict
+            return {}
+        cached_details = event.attributes.get(user_id)
+        if cached_details:
+            return cached_details
+
+        user_details = await self._send_api_request(
+            url=f"{self.api_url}/user/{user_id}"
+        )
+        event.attributes[user_id] = user_details
+        return user_details
+
+    async def get_cached_target_details(self, target_id: str) -> dict[str, Any]:
+        cached_details = event.attributes.get(target_id)
+        if cached_details:
+            return cached_details
+
+        target_details = await self._send_api_request(
+            url=f"{self.rest_api_url}/orgs/{self.org_id}/targets/{target_id}",
+            query_params={"version": f"{self.snyk_api_version}~beta"},
+        )
+        event.attributes[target_id] = target_details
+        return target_details
+
+    async def enrich_project(self, project: dict[str, Any]) -> dict[str, Any]:
```

Comment:
maybe u should load the project into a pydantic model to get the data in a more programmatic way

Reply:
I choose not to, because describing it can take me a while, and also we will need to maintain it.
LMK how urgent you feel it is